modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java

```diff
@@ -0,0 +1,178 @@
+    @Override public <T> T serviceProxy(String name, Class<? super T> svcItf, long timeout) {
+        A.notNullOrEmpty(name, "name");
+        A.notNull(svcItf, "svcItf");
+
+        return (T)Proxy.newProxyInstance(svcItf.getClassLoader(), new Class[] {svcItf},
+            new ServiceInvocationHandler<>(name, timeout, grp));
+    }
+
+    /**
+     * Gets services facade over the specified cluster group.
+     *
+     * @param grp Cluster group.
+     */
+    ClientServices withClusterGroup(ClientClusterGroupImpl grp) {
+        A.notNull(grp, "grp");
+
+        return new ClientServicesImpl(ch, marsh, grp);
+    }
+
+    /**
+     * Service invocation handler.
+     */
+    private class ServiceInvocationHandler<T> implements InvocationHandler {
+        /** Flag "Has parameter types" mask. */
+        private static final byte FLAG_PARAMETER_TYPES_MASK = 0x02;
+
+        /** Service name. */
+        private final String name;
+
+        /** Timeout. */
+        private final long timeout;
+
+        /** Cluster group. */
+        private final ClientClusterGroupImpl grp;
+
+        /**
+         * @param name Service name.
+         * @param timeout Timeout.
+         */
+        private ServiceInvocationHandler(String name, long timeout, ClientClusterGroupImpl grp) {
+            this.name = name;
+            this.timeout = timeout;
+            this.grp = grp;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            try {
+                Collection<UUID> nodeIds = grp.nodeIds();
+
+                if (nodeIds != null && nodeIds.isEmpty())
+                    throw new ClientException("Cluster group is empty.");
+
+                return ch.service(ClientOperation.SERVICE_INVOKE,
+                    req -> writeServiceInvokeRequest(req, nodeIds, method, args),
+                    res -> utils.readObject(res.in(), false)
+                );
+            }
+            catch (ClientError e) {
+                throw new ClientException(e);
+            }
+        }
+
+        /**
+         * @param ch Payload output channel.
+         */
+        private void writeServiceInvokeRequest(
+            PayloadOutputChannel ch,
+            Collection<UUID> nodeIds,
+            Method method,
+            Object[] args
+        ) {
+            ch.clientChannel().protocolCtx().checkFeatureSupported(ProtocolBitmaskFeature.SERVICE_INVOKE);
+
+            try (BinaryRawWriterEx writer = utils.createBinaryWriter(ch.out())) {
+                writer.writeString(name);
+                writer.writeByte(FLAG_PARAMETER_TYPES_MASK); // Flags.
+                writer.writeLong(timeout);
+
+                if (nodeIds == null)
+                    writer.writeInt(0);
+                else {
+                    writer.writeInt(nodeIds.size());
+
+                    for (UUID nodeId : nodeIds)
+                        writer.writeUuid(nodeId);
```

Comment:
writeUuid supports null values - it writes an additional byte which is not needed here. See `ClientClusterGroupGetNodeIdsResponse`, for example: we usually write UUIDs as 8 bytes when null does not make sense.

Reply:
On the other hand, we have UUIDs with an extra byte in compute, partition awareness, and even handshake requests/responses. I think it should be somehow unified. At least with compute until it released (both should write `nodeIds` in the same format). WDYT? 
Should we change compute as well? Or should we leave services as is?

Reply:
I would prefer to avoid redundant data. Not because 1 byte matters for performance (in this case it does not), but because it makes things confusing for the client-side code: If there is a leading byte with uuid, it means that the value is potentially null. How should the client handle nulls here? Does not make sense.

Unfortunately, it is very tempting to use `writeUuid`, so these things keep showing up. Let's not change existing code, dealing with compat is hard.

Reply:
I've fixed it.
But it seems inconsistent now to use different formats for writing the same entity for compute and services. 
Let's fix compute the same way (by another ticket of course)? It's not released yet, so there will be no compatibility issues.